src/transformation/visitors/binary-expression/compound.ts

```diff
@@ -175,13 +189,77 @@ export function transformCompoundAssignmentStatement(
             [context.transformExpression(objExpression), context.transformExpression(indexExpression)]
         );
         const accessExpression = lua.createTableIndexExpression(obj, index);
+
+        if (isSetterSkippingCompoundAssignmentOperator(operator)) {
+            return [
+                objAndIndexDeclaration,
+                ...transformSetterSkippingCompoundAssignment(context, accessExpression, operator, rhs, node),
+            ];
+        }
+
         const operatorExpression = transformBinaryOperation(context, accessExpression, right, operator, node);
         const assignStatement = lua.createAssignmentStatement(accessExpression, operatorExpression);
         return [objAndIndexDeclaration, assignStatement];
     } else {
+        if (isSetterSkippingCompoundAssignmentOperator(operator)) {
+            const luaLhs = context.transformExpression(lhs) as lua.AssignmentLeftHandSideExpression;
+            return transformSetterSkippingCompoundAssignment(context, luaLhs, operator, rhs, node);
+        }
+
         // Simple statements
         // ${left} = ${left} ${replacementOperator} ${right}
         const operatorExpression = transformBinaryOperation(context, left, right, operator, node);
         return transformAssignment(context, lhs, operatorExpression);
     }
 }
+
+/* These setter-skipping operators will not execute the setter if result does not change.
+ * x.y ||= z does NOT call the x.y setter if x.y is already true.
+ * x.y &&= z does NOT call the x.y setter if x.y is already false.
+ * x.y ??= z does NOT call the x.y setter if x.y is already not nullish.
+ */
+type SetterSkippingCompoundAssignmentOperator =
```

Comment:
After checking out [the proposal](https://github.com/tc39/proposal-logical-assignment) it seems to be intrinsic behavior for logical assignment concept, so I'd call it `LogicalAssignmentOperator`

Reply:
Actually TS already has a `type LogicalOperator = AmpersandAmpersandToken | BarBarToken`, but it does not include `??`, I propose we keep this as it is to avoid confusing naming, I did update the type to use `ts.LogicalOperator` instead of the individual tokens.